### ISSUE TITLE
AssertionError error thrown during Epoll channel cleanup

### ISSUE THREAD

**Issue by gemmellr** [netty/netty#7159]
We received a bug report [1] that our client which uses Netty 4.1.14, was failing some user tests due to throwing an assertion error:

> java.lang.AssertionError
        at io.netty.channel.epoll.EpollEventLoop.remove(EpollEventLoop.java:159)
        at io.netty.channel.epoll.AbstractEpollChannel.doDeregister(AbstractEpollChannel.java:142)
        at io.netty.channel.epoll.AbstractEpollChannel.doClose(AbstractEpollChannel.java:118)
        at io.netty.channel.epoll.AbstractEpollStreamChannel.doClose(AbstractEpollStreamChannel.java:703)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:685)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.access$700(AbstractChannel.java:419)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$5.run(AbstractChannel.java:646)
        at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:233)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:748)

Looking into the stracktrace, it seems that it is complaining the EpollEventLoop.remove() operation is not being run on the event loop thread. From the stack, it seems to be running on the GlobalEventExecutor.

Following the trace back to see how that might happen, I supposed it could be from that executor being returned here: https://github.com/netty/netty/blob/netty-4.1.14.Final/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java#L210 and used to schedule the task. If so, I also noticed the line before the executor was returned has actually already run the method throwing the assertion later when run on the executor. From the code it seemed like this would only happen with so_linger set, and indeed looking into the test further showed it does appear to be using options to set that. 

[1] https://issues.apache.org/jira/browse/QPIDJMS-314

**Comment by normanmaurer** [comment]
@gemmellr let me check

**Comment by normanmaurer** [comment]
@gemmellr fixed by https://github.com/netty/netty/pull/7162. Please check

**Comment by gemmellr** [comment]
I gave #7162 a try out from the branch and it indeed looks to fix the issue, previously the originating client test showing the problem would fail first time for me, with the updated artifact it ran successfully in a loop for some time.

I noticed while doing so that the instructions at http://netty.io/wiki/native-transports.html#building-the-linux-native-transport could be outdated, as they mention installing libgcc.i686 glibc-devel.i686 which aren't for x86_64. Separately but related, I noticed the most recent netty-transport-native-kqueue snapshots published at https://oss.sonatype.org/content/repositories/snapshots/io/netty/netty-transport-native-kqueue/4.1.16.Final-SNAPSHOT/ lack the osx-x86_64 qualified artifact, though the oldest snapshot does have it.
